### DISCUSSION TITLE
Call for thoughts on strings as images

### DISCUSSION THREAD

**Discussion by snibgo** [ImageMagick/ImageMagick#4856] [Contribute]
IM CLI contains most of the operations that we want for string handling, if we could convert strings to and from Nx1 single-channel images. For example, strlen(s) is the image width, substr(s,from,to) is "-crop", and string searching strstr(s1,s2) is roughly "compare" with "-subimage-search". Concatenating strings is "+append".

For this, we could use a new coder I will call "strimg:". It would work like this:
```
magick strimg:"Hello world" info:
```
This would create a grayscale image with 11x1 pixels. Each pixel represents one byte of the string, scaled to QuantumRange.

Encoding would be symmetrical, so:
```
magick myfile.png strimg:
```
... would take the first row of myfile.png, convert to one channel if it isn't already, scale from QuantumRange to bytes, and emit those bytes to stdout. This could be a text file:
```
magick myfile.png strimg:out.txt
```

I haven't written a "strimg:" coder, but could do. It would be quite simple. Any thoughts?

**Comment by fmw42** [comment]
Can you explain an application for this?

**Comment by snibgo** [comment]
Suppose we want to know if an image x.png has colorspace "sRGB":
```
magick ^
  x.png ^
  strimg:%[colorspace] ^
  strimg:"sRGB" ^
  -delete 0 ^
  -compose Difference -composite ^
  -format IsSRGB=%[fx:maxima==0]
  info:
```
The output would be "IsSRGB=0" or "IsSRGB=1".

The string processing could be done externally to IM, with a shell script. But if we want conditional processing that depends on the image colorspace, doing all the work within IM has performance benefits.

**Reply by fmw42** [reply]
Thanks. That helps explain the benefits and usage.

**Comment by snibgo** [comment]
"Strings as images" is a small helper towards a larger project. The example above is quite clumsy, but we could do the string processing in a function and use that like this:
```
magick ^
  x.png ^
  if IsSRGB() then
    -print "%f is already sRGB"
  else
    -colorspace sRGB
  endif
  -write out.png
```
Functions (with parameters and local variables), "if...then", "while...do" and so on are under development.

**Reply by dlemstra** [reply]
Not sure if we should move all the scripting inside ImageMagick. It will have some performance benefits but that will also introduce a lot of complexity in the library.

**Comment by snibgo** [comment]
It doesn't need to add any complexity to existing code. But that is a separate topic.

Any thoughts on "strings as images"?

**Comment by dlemstra** [comment]
I have read your example and message multiple times and I don't understand what this coder does and how it would help someone. 

And I was talking about this syntax:
```
  if IsSRGB() then
    -print "%f is already sRGB"
  else
    -colorspace sRGB
  endif
```

This looks like something that should not be part of ImageMagick itself.

**Comment by snibgo** [comment]
The decoder "strimg:" would read text and make an image file. The encoder "strimg:" would read an image file and write text.

It would enable string operations to be performed within IM, with no other changes needed to IM.

**Comment by dlemstra** [comment]
I am sorry but I don't understand why that would enable string operations inside IM and I also don't understand why we would need that.

**Comment by urban-warrior** [comment]
We added support for the STRIMG format to the main branch.  Test it and let us know if there are any issues.

**Comment by snibgo** [comment]
Thanks. This works as expected for literal strings, and for strings from "@" files:
```
magick strimg:"abc" info:
strimg:abc STRIMG 3x1 3x1+0+0 8-bit Gray 0.718u 0:00.725

magick strimg:"abc" strimg:
abc

magick strimg:@textfile.txt strimg:
This
Is
A
Test
```

This also works, but takes 22 seconds:
```
magick rose: -colorspace CMYK -set option:colsp %[colorspace] strimg:%[colsp] -delete 0 strimg:
CMYK
```
I don't know why it takes so long. It may be a a build problem at my end.

**Comment by urban-warrior** [comment]
We'll fix the speed issue.  Look for an update in the main branch soon.

**Comment by mviereck** [comment]
I am not sure if this fits exactly here, but at least it is related.
I've tried to write a NetPBM image literally on command line, but was forced to use a base64 encoded string instead:
https://github.com/ImageMagick/ImageMagick/discussions/5707

`strimg:` might be a way to allow a literal NetPBM image on cli. However, I did not understand `strimg:` enough to estimate that.
